api/authentication.py

```diff
@@ -10,77 +10,101 @@
-        # the 'exp' time returned by FXA
+        # the 'exp' time in the JWT returned by FXA
         cache_timeout = 60
         if not fxa_resp_data:
             introspect_token_url = (
                 "%s/introspect"
                 % settings.SOCIALACCOUNT_PROVIDERS["fxa"]["OAUTH_ENDPOINT"]
             )
-            fxa_resp = requests.post(introspect_token_url, json={"token": token})
-            fxa_resp_data = {"status_code": fxa_resp.status_code, "json": None}
-            try:
-                fxa_resp_data["json"] = fxa_resp.json()
-            except requests.exceptions.JSONDecodeError:
-                logger.error(
-                    "JSONDecodeError from FXA introspect response.",
-                    extra={"fxa_response": shlex.quote(fxa_resp.text)},
-                )
+            fxa_resp_data = introspect_token(
+                introspect_token_url, token, cache_key, cache_timeout
+            )
 
         user = None
-        if (
-            fxa_resp_data
-            and fxa_resp_data["status_code"] == 200
-            and fxa_resp_data["json"]
-            and fxa_resp_data["json"].get("active")
-        ):
-
-            # FxA user is active, check for the associated Relay account
-            fxa_uid = fxa_resp_data.get("json").get("sub")
-            if fxa_uid:
-                try:
-                    sa = SocialAccount.objects.get(uid=fxa_uid, provider="fxa")
-                except SocialAccount.DoesNotExist:
-                    # No Relay account associated with the FxA ID. It might be
-                    # a user who deleted their Relay account since they first
-                    # signed up
-                    pass
-                else:
-                    user = sa.user
-
-                    # cache fxa_resp_data for as long as access_token is valid
-                    # Note: FXA iat and exp are timestamps in *milliseconds*
-                    fxa_token_exp_time = int(
-                        fxa_resp_data.get("json").get("exp") / 1000
-                    )
-                    now_time = int(datetime.now(timezone.utc).timestamp())
-                    cache_timeout = fxa_token_exp_time - now_time
+        if not fxa_resp_data["status_code"] == 200:
+            raise AuthenticationFailed("Did not receive a 200 response from FXA.")
+
+        if not fxa_resp_data["json"].get("active"):
+            raise AuthenticationFailed("FXA returned active: False for token.")
+
+        # FxA user is active, check for the associated Relay account
+        fxa_uid = fxa_resp_data.get("json").get("sub")
+        if not fxa_uid:
+            raise AuthenticationFailed(
+                "Authenticated user does not have a Relay account."
+            )
+        try:
+            sa = SocialAccount.objects.get(uid=fxa_uid, provider="fxa")
+        except SocialAccount.DoesNotExist:
+            raise AuthenticationFailed(
+                "Authenticated user does not have a Relay account."
```

Comment:
Nit: should this message be different than the one if there is no `fxa_uid`?

Reply:
Updated the previous one to "FXA did not return an FXA UID."